readthedocs/projects/tasks/search.py

```diff
@@ -40,112 +42,215 @@ def fileify(version_pk, commit, build, search_ranking, search_ignore):
+        return
+
+    log.bind(
+        project_slug=version.project.slug,
+        version_slug=version.slug,
+        build_id=build.id,
+    )
+
+    search_ranking = []
+    search_ignore = []
+    build_config = build.config
+    if build_config:
+        search_config = build_config.get("search", {})
+        search_ranking = search_config.get("ranking", [])
+        search_ignore = search_config.get("ignore", [])
+
     try:
-        _sync_imported_files(version, build)
+        _create_imported_files_and_search_index(
+            version=version,
+            build_id=build.id,
+            search_ranking=search_ranking,
+            search_ignore=search_ignore,
+        )
     except Exception:
-        log.exception('Failed during ImportedFile syncing')
+        log.exception("Failed during creation of new files")
 
 
-def _sync_imported_files(version, build):
+@app.task(queue="reindex")
+def reindex_version(version_id, search_index_name=None):
     """
-    Sync/Update/Delete ImportedFiles objects of this version.
+    Re-create imported files and search index for the version.
 
-    :param version: Version instance
-    :param build: Build id
+    The latest successful build is used for the re-creation.
     """
-    project = version.project
+    version = Version.objects.filter(pk=version_id).select_related("project").first()
+    if not version or not version.built:
+        return
 
-    # Index new HTMLFiles to ElasticSearch
-    index_new_files(model=HTMLFile, version=version, build=build)
+    latest_successful_build = (
+        version.builds.filter(state=BUILD_STATE_FINISHED, success=True)
+        .order_by("-date")
+        .first()
+    )
+    # If the version doesn't have a successful
+    # build, we don't have files to index.
+    if not latest_successful_build:
+        return
 
-    # Remove old HTMLFiles from ElasticSearch
-    remove_indexed_files(
-        model=HTMLFile,
+    log.bind(
         project_slug=version.project.slug,
         version_slug=version.slug,
-        build_id=build,
+        build_id=latest_successful_build.id,
     )
 
-    # Delete ImportedFiles objects (including HTMLFiles)
-    # from the previous build of the version.
-    (
-        ImportedFile.objects.filter(project=project, version=version)
-        .exclude(build=build)
-        .delete()
-    )
+    search_ranking = []
+    search_ignore = []
+    build_config = latest_successful_build.config
+    if build_config:
+        search_config = build_config.get("search", {})
+        search_ranking = search_config.get("ranking", [])
+        search_ignore = search_config.get("ignore", [])
 
+    # We need to use a build id that is different from the current one,
+    # otherwise we will end up with duplicated files. After the process
+    # is complete, we delete the files and search index that don't belong
+    # to the current build id. The build id isn't used for anything else.
+    build_id = latest_successful_build.id + 1
```

Comment:
This looks pretty hacky. I don't really understand what's going on here and why we are using a `build.id` that may not even belong to the current project.

Can you expand on the explanation of what's happening behind the scenes here?

If we _require_ an exceptional use case of the `_create_imported_files_and_search_index` I'd like to make that pretty explicit on the function definition instead of passing a random/fake `build.id` that's not the real build id. Otherwise, this will lead to pretty confusing scenarios in the near future.

Reply:
Build id can be anything, it just needs to be different from the current one we have in the search index. I'm not renaming it, since that will some planning. I'm going to change it, so the generation of the ID happens inside the function itself, and will create an issue to rename this to sync_id.

Reply:
https://github.com/readthedocs/readthedocs.org/pull/10696/commits/8fdbc0681f05c5706332f65d2e27986dc60df13b

Reply:
Perfect! This is a lot more clear now 👍🏼